src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -413,8 +503,149 @@ fn compute_deps_custom_build(
         // Build scripts always compiled for the host.
         CompileKind::Host,
         CompileMode::Build,
+        /*artifact*/ false,
     )?;
-    Ok(vec![unit_dep])
+
+    let artifact_build_deps = state.deps(unit, script_unit_for, &|dep| {
+        dep.kind() == DepKind::Build && dep.artifact().is_some()
+    });
+
+    if artifact_build_deps.is_empty() {
+        Ok(vec![compile_script_unit])
+    } else {
+        let mut artifact_units: Vec<_> = build_artifact_requirements_to_units(
+            unit,
+            unit_for.root_compile_kind(),
+            artifact_build_deps,
+            state,
+        )?;
+        artifact_units.push(compile_script_unit);
+        Ok(artifact_units)
+    }
+}
+
+fn build_artifact_requirements_to_units(
+    parent: &Unit,
+    root_unit_compile_target: CompileKind,
+    artifact_deps: Vec<(PackageId, &HashSet<Dependency>)>,
+    state: &State<'_, '_>,
+) -> CargoResult<Vec<UnitDep>> {
+    let mut ret = Vec::new();
+    // So, this really wants to be true for build dependencies, otherwise resolver = "2" will fail.
+    // It means that the host features will be separated from normal features, thus won't be unified with them.
+    let host_features = true;
+    let unit_for = UnitFor::new_host(host_features, root_unit_compile_target);
+    for (dep_pkg_id, deps) in artifact_deps {
+        let artifact_pkg = state.get(dep_pkg_id);
+        for build_dep in deps.iter().filter(|d| d.is_build()) {
+            ret.extend(artifact_targets_to_unit_deps(
+                parent,
+                unit_for,
+                state,
+                build_dep
+                    .artifact()
+                    .expect("artifact dep")
+                    .target()
+                    .map(|kind| match kind {
+                        ArtifactTarget::Force(target) => CompileKind::Target(target),
+                        ArtifactTarget::BuildDependencyAssumeTarget => root_unit_compile_target,
+                    })
+                    .unwrap_or(CompileKind::Host),
+                artifact_pkg,
+                build_dep,
+            )?);
+        }
+    }
+    Ok(ret)
+}
+
+fn artifact_targets_to_unit_deps(
+    parent: &Unit,
+    parent_unit_for: UnitFor,
+    state: &State<'_, '_>,
+    compile_kind: CompileKind,
+    artifact_pkg: &Package,
+    dep: &Dependency,
+) -> CargoResult<Vec<UnitDep>> {
+    let ret = match_artifacts_kind_with_targets(parent, dep, artifact_pkg.targets())?
+        .into_iter()
+        .flat_map(|target| {
+            // We split target libraries into individual units, even though rustc is able to produce multiple
+            // kinds in an single invocation for the sole reason that each artifact kind has its own output directory,
+            // something we can't easily teach rustc for now.
```

Comment:
> In theory, if someone happened to build a Rust library and a cdylib and a staticlib, those might be able to get emitted by the same rustc invocation, but that seems like a much less common case, and if we want to optimize for it at all, we could do that later.

> That said, we might still want to invoke rustc with a different out directory (e.g. the deps directory, since the hash should be unique), then use cargo's existing knowledge of rustc output files (used elsewhere to "uplift" the final binaries) to copy them to the right place.

Thanks for bringing this up. To add context, splitting the dependency unit into multiple adds quite some complexity as the return value is now a `Vec`, even though most of the time (i.e builds without artifact deps) it will only contain a single unit. I have noticed the 'uplift' capability as well and believe this could be used to achieve the directory split as well.

Since Rustc supports it, I think it would be wrong ignoring that capability here and suggest to turn it into an 'uplift' based mechanism.

Once you give the go ahead it will be done.

Reply:
My understanding is that you would *still* need multiple rustc invocations for cases like `artifact = ["bin"], lib = true`, or `artifact = ["bin"]` with multiple binaries. You already need to return a `Vec` for those cases.

I think there's value in doing those rustc invocations in a separate directory, and then uplifting the binaries to the appropriate directory for the artifact type. But don't worry about trying to invoke rustc fewer times, since there will still be cases where you *have* to invoke rustc multiple times.

Reply:
Also, note that doing those rustc invocations in a separate directory does not necessarily need to happen in *this* PR.

Reply:
That's right, `Vec` is indeed required and there is probably no need to optimize the amount of times `rustc` is called (_just yet_).

I am happy to leave the uplifting of build results to `RFC-3176`, a review note was added there so this won't be forgotten.